### PR TITLE
#2599 DynaFlow : add Rpcl2 model + nrt

### PR DESCRIPTION
**Pull request by cossartque**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

@rosiereflo fyi

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+model ReactivePowerControlLoop2 "Simplified Reactive Power Control Loop model for renewables"
+  /* The reactive control loop gets a level K from the secondary voltage control and transforms it to a voltage reference for the generator voltage regulator */
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:16]: This comment would better be in the documentation.

Also, "transforms it into a voltage reference".

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  import Modelica;
+  import Dynawo.Types;
+
+  parameter Types.PerUnit QrPu "Participation factor of the generator to the secondary voltage control in pu (base QNom)";
+  parameter Types.PerUnit DeltaURefMaxPu "Maximum of deltaUref on one activation period (Uref(t+1) - Uref(t)) (base UNom)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:21]: Maximum of deltaU**R**ef on one activation period (U**R**ef(t+1) - U**R**ef(t)) **in pu** (base UNom)

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  parameter Types.PerUnit CqMaxPu "Max of Cq in the different exploitation schemes in pu (base UNom, QNom)";
+  parameter Types.Time Ti "Filters' time constant in s";
+  parameter Types.Time Tech "Sampling time in s";
+  parameter Types.Time Tech2 = Tech "Integrator's time constant equal to sampling time in s";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:25]: tSample and tSample2

Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:25]: Please respect the alphabetical order and start the time parameters with t.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    Placement(visible = true, transformation(origin = {-40, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.FirstOrder firstOrder1(T = Ti)  annotation(
+    Placement(visible = true, transformation(origin = {50, -40}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+  Modelica.Blocks.Math.Gain gain(k = Ti / Tech)  annotation(
+    Placement(visible = true, transformation(origin = {10, -40}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+
+  parameter Types.PerUnit UStatorRef0Pu "Start value of the generator stator voltage reference in pu (base UNom)";
+  parameter Types.PerUnit QStator0Pu "Start value of the generator stator reactive power in pu (base QNomAlt) (generator convention)";
+  parameter Boolean limUQUp0 "Whether the maximum reactive power limits are reached or not (from generator voltage regulator), start value";
+  parameter Boolean limUQDown0 "Whether the minimum reactive power limits are reached or not (from generator voltage regulator), start value";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:76]: Please respect the alphabetical order

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  // Blocks
+  Modelica.Blocks.Math.Gain participation(k = QrPu) annotation(
+    Placement(visible = true, transformation(origin = {-130, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Math.Feedback errQ annotation(
+    Placement(visible = true, transformation(origin = {-100, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.Integrator integrator(k = 1 / Tech2, y_start = UStatorRef0Pu) annotation(
+    Placement(visible = true, transformation(origin = {110, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Math.Gain gainIntegrator(k = 1 / CqMaxPu) annotation(
+    Placement(visible = true, transformation(origin = {-70, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Nonlinear.VariableLimiter rampLim(limitsAtInit = true) annotation(
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:50]: (limitsAtInit = true) is useless and can be removed.

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  parameter Types.Time Tech "Sampling time in s";
+  parameter Types.Time Tech2 = Tech "Integrator's time constant equal to sampling time in s";
+
+  // Input variables
+  Modelica.Blocks.Interfaces.RealInput level "Level received from the secondary voltage control [-1;1] " annotation(
+    Placement(visible = true, transformation(origin = {-170, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {-160, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.BooleanInput limUQUp(start = limUQUp0) "Whether the maximum reactive power limits are reached or not" annotation(
+    Placement(visible = true, transformation(origin = {-170, 80}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {-160, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Interfaces.BooleanInput limUQDown(start = limUQDown0) "Whether the minimum reactive power limits are reached or not" annotation(
+    Placement(visible = true, transformation(origin = {-170, -120}, extent = {{-10, -10}, {10, 10}}, rotation = 0), iconTransformation(origin = {-160, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:33]: Alphabetical order : limUQUp after limUQDown

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo (hunk 3 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:25]: Ok for order, I keep the parameters for DB reasons and consistency with DynaWaltz

Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:25]: Understood

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo [hunk 7]
```diff
@@ -59,10 +58,10 @@ model ReactivePowerControlLoop "Simplified Reactive Power Control Loop model"
   Modelica.Blocks.Sources.Constant const3(k = -DerURefMaxPu) annotation(
     Placement(visible = true, transformation(origin = {-170, -160}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
 
-  parameter Types.PerUnit UStatorRef0Pu "Start value of the generator stator voltage reference in pu (base UNom)";
-  parameter Types.PerUnit QStator0Pu "Start value of the generator stator reactive power in pu (base QNomAlt) (generator convention)";
-  parameter Boolean limUQUp0 "Whether the maximum reactive power limits are reached or not (from generator voltage regulator), start value";
   parameter Boolean limUQDown0 "Whether the minimum reactive power limits are reached or not (from generator voltage regulator), start value";
+  parameter Boolean limUQUp0 "Whether the maximum reactive power limits are reached or not (from generator voltage regulator), start value";
+  parameter Types.PerUnit QStator0Pu "Start value of the generator stator reactive power in pu (base QNomAlt) (generator convention)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo:63]: Types.ReactivePowerPu

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo [hunk 8]
```diff
@@ -59,10 +58,10 @@ model ReactivePowerControlLoop "Simplified Reactive Power Control Loop model"
...
+  parameter Types.PerUnit UStatorRef0Pu "Start value of the generator stator voltage reference in pu (base UNom)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo:64]: Types.VoltageModulePu

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo [hunk 9]
```diff
@@ -13,21 +13,20 @@ within Dynawo.Electrical.Controls.Machines.ReactivePowerControlLoops;
 */
 
 model ReactivePowerControlLoop "Simplified Reactive Power Control Loop model"
-  /* The reactive control loop gets a level K from the secondary voltage control and transforms it to a voltage reference for the generator voltage regulator */
   import Modelica;
   import Dynawo.Types;
 
+  parameter Types.PerUnit DerURefMaxPu "Maximum variation rate of UStatorRefPu in pu (base UNom)";
   parameter Types.PerUnit QrPu "Participation factor of the generator to the secondary voltage control in pu (base QNomAlt)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo:20]: Types.ReactivePowerPu

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo [hunk 10]
```diff
@@ -13,21 +13,20 @@ within Dynawo.Electrical.Controls.Machines.ReactivePowerControlLoops;
 */
 
 model ReactivePowerControlLoop "Simplified Reactive Power Control Loop model"
-  /* The reactive control loop gets a level K from the secondary voltage control and transforms it to a voltage reference for the generator voltage regulator */
   import Modelica;
   import Dynawo.Types;
 
+  parameter Types.PerUnit DerURefMaxPu "Maximum variation rate of UStatorRefPu in pu (base UNom)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop.mo:19]: in pu/s (base UNom)

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  Modelica.Blocks.Math.Feedback feedback annotation(
+    Placement(visible = true, transformation(origin = {-40, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
+  Modelica.Blocks.Continuous.FirstOrder firstOrder1(T = Ti)  annotation(
+    Placement(visible = true, transformation(origin = {50, -40}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+  Modelica.Blocks.Math.Gain gain(k = Ti / Tech)  annotation(
+    Placement(visible = true, transformation(origin = {10, -40}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
+
+  parameter Boolean limUQDown0 "Whether the minimum reactive power limits are reached or not (from generator voltage regulator), start value";
+  parameter Boolean limUQUp0 "Whether the maximum reactive power limits are reached or not (from generator voltage regulator), start value";
+  parameter Types.PerUnit QStator0Pu "Start value of the generator stator reactive power in pu (base QNomAlt) (generator convention)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:74]: Types.ReactivePowerPu

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  parameter Types.PerUnit UStatorRef0Pu "Start value of the generator stator voltage reference in pu (base UNom)";
```
Comment (ErwanGuichard49) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:75]: Types.VoltageModulePu

**Review by ErwanGuichard49 — DISMISSED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+* All rights reserved.
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools for power systems.
+*/
+
+model ReactivePowerControlLoop2 "Simplified Reactive Power Control Loop model for renewables"
```
Comment (marcochiaramello) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:15]: renewables energy sources

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+  import Modelica;
+  import Dynawo.Types;
+
+  parameter Types.PerUnit CqMaxPu "Max of Cq in the different exploitation schemes in pu (base UNom, QNom)";
+  parameter Types.PerUnit DeltaURefMaxPu "Maximum of deltaURef on one activation period (URef(t+1) - URef(t)) in pu (base UNom)";
```
Comment (marcochiaramello) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo:20]: Types.VoltageModulePu

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo (hunk 13 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Machines/ReactivePowerControlLoops/ReactivePowerControlLoop2.mo]: renewable energy sources

**Pull request merged**